skrub/_expressions/_subsampling.py

```diff
@@ -0,0 +1,139 @@
+import numpy as np
+
+from .. import _dataframe as sbd
+from . import _evaluation, _expressions
+
+SHOULD_SUBSAMPLE_KEY = "_skrub_should_subsample"
+
+
+def _should_subsample(mode, environment):
+    if mode == "preview":
+        return True
+    if "fit" not in mode:
+        return False
+    return environment.get(SHOULD_SUBSAMPLE_KEY, False)
+
+
+class ShouldSubsample(_expressions.ExprImpl):
+    _fields = []
+
+    def compute(self, e, mode, environment):
+        return _should_subsample(mode, environment)
+
+
+@_expressions.check_expr
+def should_subsample():
+    """Expression indicating if subsampling should be applied.
+
+    This is a helper for other skrub functions that need the information of
+    whether they should apply any subsampling, depending on the current
+    evaluation mode and parameters passed by the user.
+
+    Examples
+    --------
+    >>> import skrub
+    >>> from skrub._expressions._subsampling import should_subsample
+
+    >>> @skrub.deferred
+    ... def load_data(subsampling=should_subsample()):
+    ...     print("subsampling:", subsampling)
+    ...     return 1.0 if subsampling else 2.0
+
+    >>> e = load_data()
+    subsampling: True
+    >>> e
+    <Call 'load_data'>
+    Result (on a subsample):
+    ――――――――――――――――――――――――
+    1.0
+    >>> e.skb.get_pipeline(fitted=True)
+    subsampling: False
+    SkrubPipeline(expr=<Call 'load_data'>)
+    >>> e.skb.get_pipeline(keep_subsampling=True, fitted=True)
+    subsampling: True
+    SkrubPipeline(expr=<Call 'load_data'>)
+    """
+    return _expressions.Expr(ShouldSubsample())
+
+
+def _sample_numpy(a, n):
```

Comment:
Shouldn't we have a way to control this sampling stream?

I would probably advise for using a dedicated random number generator (not seeded by default), and documenting somewhere how to control it. Some users will want to do that.

That comment applies to the whole subsampling code

Reply:
I think we want it to be seeded by default so that if you sample X and y separately you get the same indices and don't break their alignment. We can expose the seed parameter though, I just figured few people would use it

Reply:
@GaelVaroquaux  I think this is the last outstanding comment. do you want to expose the seed as a parameter or keep it hard-coded to 0?

Reply:
I don't want this seeded by default.

I would like a way for the user to globally control the seed / random number sequence for subsampling.

The way I would consider doing it is by having a dedicated rng, exposed as a global somewhere. But I'm open to other implementations of a solution.

Reply:
but if it's not seeded don't you think some users will shoot themselves in the foot by doing `X.skb.subsample(how='random')` and later `y.skb.subsample(how='random')` and ending up selecting different indices in X and y and breaking the correspondance between the 2?

Reply:
Point taken.

OK, I think that a solution for now, without over-engineering, would be to expose a default "SUBSAMPLING_SEED" and use it.

Drawback is that it pollutes the skrub namespace. That would be an argument for giving up on my idea. As you wish

Reply:
Skrub config is a good idea

Reply:
Let's add this to an issue and move on